File: VB/TreeListFilter/FilterTreeList/ColumnFilter/ColumnFilterItem.vb
Imports System
Imports DevExpress.XtraTreeList.Columns
Imports DevExpress.XtraTreeList.Nodes

Namespace FilterTreeListControl
	Public Class ColumnFilterItem
		Private ReadOnly _value As Object
        Private ReadOnly _displayText As String

		Public Sub New(ByVal value As Object, ByVal displayText As String)
			Me._value = value
			Me._displayText = displayText
		End Sub
		Public Sub New(ByVal node As TreeListNode, ByVal col As TreeListColumn)
			_value = node.GetValue(col)
			_displayText = node.GetDisplayText(col)
		End Sub

		Protected Overridable Function GetItemIsEmpty() As Boolean
			Return (_value Is Nothing OrElse _value Is DBNull.Value)
		End Function

		Public Overrides Overloads Function Equals(ByVal obj As Object) As Boolean
			If Not(TypeOf obj Is ColumnFilterItem) Then
				Return MyBase.Equals(obj)
			End If

			Dim item As ColumnFilterItem = CType(obj, ColumnFilterItem)
			Return Object.Equals(Me.Value, item.Value) AndAlso Me.DisplayText = item.DisplayText
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return MyBase.GetHashCode()
		End Function

		Public Shared Operator =(ByVal item1 As ColumnFilterItem, ByVal item2 As ColumnFilterItem) As Boolean
			If Object.Equals(item1, Nothing) AndAlso Object.Equals(item2, Nothing) Then
				Return True
			End If

			If Object.Equals(item1, Nothing) AndAlso (Not Object.Equals(item2, Nothing)) Then
				Return False
			End If

			If (Not Object.Equals(item1, Nothing)) AndAlso Object.Equals(item2, Nothing) Then
				Return False
			End If

			Return item1.Equals(item2)
		End Operator

		Public Shared Operator <>(ByVal item1 As ColumnFilterItem, ByVal item2 As ColumnFilterItem) As Boolean
			Return Not(item1 = item2)
		End Operator

		Public Overrides Function ToString() As String
			If DisplayText Is Nothing Then
				Return String.Empty
			End If

			Return DisplayText
		End Function

		Public ReadOnly Property Value() As Object
			Get
				Return _value
			End Get
		End Property

		Public ReadOnly Property DisplayText() As String
			Get
				Return _displayText
			End Get
		End Property

		Public ReadOnly Property IsEmpty() As Boolean
			Get
				Return GetItemIsEmpty()
			End Get
		End Property
	End Class

	Public Class ColumnFilterCustomItem
		Inherits ColumnFilterItem
		Public Sub New()
			MyBase.New("(Custom)", "(Custom)")
		End Sub
	End Class

	Public Class ColumnFilterAllItem
		Inherits ColumnFilterItem
		Public Sub New()
			MyBase.New("(All)", "(All)")
		End Sub
	End Class

	Public Class ColumnFilterBlanksItem
		Inherits ColumnFilterItem
		Public Sub New()
			MyBase.New("(Blanks)", "(Blanks)")
		End Sub
	End Class

	Public Class ColumnFilterNonBlanksItem
		Inherits ColumnFilterItem
		Public Sub New()
			MyBase.New("(Non blanks)", "(Non blanks)")
		End Sub
	End Class
End Namespace

File: CS/TreeListFilter/FilterTreeList/ColumnFilter/ColumnFilterItem.cs
using System;
using DevExpress.XtraTreeList.Columns;
using DevExpress.XtraTreeList.Nodes;

namespace FilterTreeListControl
{
	public class ColumnFilterItem
	{
		private readonly object value;
		private readonly string displayText;

		public ColumnFilterItem(object value, string displayText)
		{
			this.value = value;
			this.displayText = displayText;
		}
		public ColumnFilterItem(TreeListNode node, TreeListColumn col)
		{
			value = node.GetValue(col);
			displayText = node.GetDisplayText(col);
		}

		protected virtual bool GetItemIsEmpty()
		{
			return (value == null || value == DBNull.Value);
		}

		public override bool Equals(object obj)
		{
			if ( !(obj is ColumnFilterItem) )
				return base.Equals(obj);

			ColumnFilterItem item = (ColumnFilterItem)obj;
			return Object.Equals(this.Value, item.Value) && this.DisplayText == item.DisplayText;
		}

		public override int GetHashCode()
		{
			return base.GetHashCode();
		}

		public static bool operator ==(ColumnFilterItem item1, ColumnFilterItem item2)
		{
			if ( Object.Equals(item1, null) && Object.Equals(item2, null) )
				return true;

			if ( Object.Equals(item1, null) && !Object.Equals(item2, null) )
				return false;

			if ( !Object.Equals(item1, null) && Object.Equals(item2, null) )
				return false;

			return item1.Equals(item2);
		}

		public static bool operator !=(ColumnFilterItem item1, ColumnFilterItem item2)
		{
			return !(item1 == item2);
		}

		public override string ToString()
		{
			if ( DisplayText == null )
				return String.Empty;

			return DisplayText;
		}

		public object Value
		{
			get { return value; }
		}

		public string DisplayText
		{
			get { return displayText; }
		}

		public bool IsEmpty
		{
			get { return GetItemIsEmpty(); }
		}
	}

	public class ColumnFilterCustomItem : ColumnFilterItem
	{
		public ColumnFilterCustomItem()
			: base("(Custom)", "(Custom)")
		{
		}
	}

	public class ColumnFilterAllItem : ColumnFilterItem
	{
		public ColumnFilterAllItem()
			: base("(All)", "(All)")
		{
		}
	}

	public class ColumnFilterBlanksItem : ColumnFilterItem
	{
		public ColumnFilterBlanksItem()
			: base("(Blanks)", "(Blanks)")
		{
		}
	}

	public class ColumnFilterNonBlanksItem : ColumnFilterItem
	{
		public ColumnFilterNonBlanksItem()
			: base("(Non blanks)", "(Non blanks)")
		{
		}
	}
}
